src/versionbits.cpp

```diff
@@ -74,12 +74,16 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
                 }
                 if (count >= nThreshold) {
                     stateNext = ThresholdState::LOCKED_IN;
+                } else if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
+                    stateNext = ThresholdState::FAILED;
                 }
                 break;
             }
             case ThresholdState::LOCKED_IN: {
-                // Always progresses into ACTIVE.
-                stateNext = ThresholdState::ACTIVE;
+                // Progresses into ACTIVE provided activation height will have been reached.
+                if (pindexPrev->nHeight + 1 >= min_activation_height) {
```

Comment:
nit: Might be worth adding a comment saying `pindexPrev->nHeight + 1` is used so we measure the current block height, and not the previous block's height

Reply:
I :+1: the sentiment that it's confusing, but think it's clear enough in this case.

In #21392 there were a lot of examples of `pindexPrev->nHeight + 1 < min_activation_height` which I wanted documented better. Despite being the negation, is a lot harder to understand when the transition was happening because it was in states with multiple possible transitions, and the clause was guarding a state transition that becomes *impossible* rather than one that becomes *possible*. As such, even if this code contained an off-by-one error in the logic, it would only result in one additional period of LOCKED_IN as opposed to a failure to activate.

That said, there's no off-by-one error here AFAICT :)